File: bindings/kotlin/src/main/kotlin/at/schem/nucleation/Field3.kt
package at.schem.nucleation;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface Field3Lib: Library {
    fun Field3_destroy(handle: Pointer)
    fun Field3_value_noise_fbm(frequency: Float, seed: Int, octaves: FFIUint32): ResultPointerInt
    fun Field3_eval_at(handle: Pointer, x: Float, y: Float, z: Float): Float
    fun Field3_output_range(handle: Pointer): ResultFieldRangeNativeInt
    fun Field3_from_json_string(json: Slice): ResultPointerInt
    fun Field3_to_json(handle: Pointer, write: Pointer): ResultUnitInt
}
/** Immutable scalar field evaluated over world-space `(x, y, z)`.
*
*A `Field3` has scalar semantics only. It may be shared by geometry and
*material consumers without being reinterpreted as a signed-distance field.
*/
class Field3 internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal var owned: Boolean,
)  {

    init {
        if (this.owned) {
            this.registerCleaner()
        }
    }

    private class Field3Cleaner(val handle: Pointer, val lib: Field3Lib) : Runnable {
        override fun run() {
            lib.Field3_destroy(handle)
        }
    }
    private fun registerCleaner() {
        CLEANER.register(this, Field3.Field3Cleaner(handle, Field3.lib));
    }

    companion object {
        internal val libClass: Class<Field3Lib> = Field3Lib::class.java
        internal val lib: Field3Lib = Native.load("nucleation", libClass)
        @JvmStatic

        /** Deterministic value-noise FBM normalized to `[-1, 1]`.
        */
        fun valueNoiseFbm(frequency: Float, seed: Int, octaves: UInt): Result<Field3> {

            val returnVal = lib.Field3_value_noise_fbm(frequency, seed, FFIUint32(octaves));
            val nativeOkVal = returnVal.getNativeOk();
            if (nativeOkVal != null) {
                val selfEdges: List<Any> = listOf()
                val handle = nativeOkVal
                val returnOpaque = Field3(handle, selfEdges, true)
                return returnOpaque.ok()
            } else {
                return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
            }
        }
        @JvmStatic

        fun fromJsonString(json: String): Result<Field3> {
            val jsonSliceMemory = PrimitiveArrayTools.borrowUtf8(json)

            val returnVal = lib.Field3_from_json_string(jsonSliceMemory.slice);
            try {
                val nativeOkVal = returnVal.getNativeOk();
                if (nativeOkVal != null) {
                    val selfEdges: List<Any> = listOf()
                    val handle = nativeOkVal
                    val returnOpaque = Field3(handle, selfEdges, true)
                    return returnOpaque.ok()
                } else {
                    return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
                }
            } finally {
                jsonSliceMemory.close()
            }
        }
    }

    fun evalAt(x: Float, y: Float, z: Float): Float {

        val returnVal = lib.Field3_eval_at(handle, x, y, z);
        return (returnVal)
    }

    /** The field's analytically proven output range.
    *
    *Returns `NotFound` when no range can be proven — callers mapping a
    *field onto a gradient must handle that rather than silently
    *propagating a sentinel into their `lo`/`hi` bounds.
    */
    fun outputRange(): Result<FieldRange> {

        val returnVal = lib.Field3_output_range(handle);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {
            val returnStruct = FieldRange.fromNative(nativeOkVal)
            return returnStruct.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }

    fun toJson(): Result<String> {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Field3_to_json(handle, write);
        val nativeOkVal = returnVal.getNativeOk();
        if (nativeOkVal != null) {

            val returnString = DW.writeToString(write)
            return returnString.ok()
        } else {
            return NucleationErrorError(NucleationError.fromNative(returnVal.getNativeErr()!!)).err()
        }
    }

}

File: bindings/php/Field3.php
<?php
namespace Stencil;

final class Field3 {
    /** @internal */
    public \FFI\CData $ptr;
    private bool $owned;
    private ?object $borrowedFrom;

    /** @internal */
    public function __construct(\FFI\CData $ptr, bool $owned, ?object $borrowedFrom = null) {
        $this->ptr = $ptr;
        $this->owned = $owned;
        $this->borrowedFrom = $borrowedFrom;
    }

    public static function valueNoiseFbm( $frequency,  $seed,  $octaves) {
        $result = Lib::ffi()->Field3_value_noise_fbm($frequency, $seed, $octaves);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new Field3($result->ok, true);
    }

    public function evalAt( $x,  $y,  $z) {
        $ret = Lib::ffi()->Field3_eval_at($this->ptr, $x, $y, $z);
        return $ret;
    }

    public function outputRange() {
        $result = Lib::ffi()->Field3_output_range($this->ptr);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return FieldRange::fromFFI($result->ok);
    }

    public static function fromJsonString(string $json) {
        $__n0 = strlen($json);
        $__view0 = Lib::ffi()->new('DiplomatStringView');
        if ($__n0 > 0) {
            $__buf0 = Lib::ffi()->new("uint8_t[" . $__n0 . "]", false);
            \FFI::memcpy($__buf0, $json, $__n0);
            $__view0->data = Lib::ffi()->cast('const char*', \FFI::addr($__buf0[0]));
        } else {
            $__view0->data = null;
        }
        $__view0->len = $__n0;
        $result = Lib::ffi()->Field3_from_json_string($__view0);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return new Field3($result->ok, true);
    }

    public function toJson() {
        $write = Lib::ffi()->diplomat_buffer_write_create(0);
        $result = Lib::ffi()->Field3_to_json($this->ptr, $write);
        if (!$result->is_ok) {
            throw new DiplomatError('NucleationError', $result->err, NucleationError::name($result->err));
        }
        return Lib::readAndFreeWrite($write);
    }

    public function __destruct() {
        if ($this->owned) {
            Lib::ffi()->Field3_destroy($this->ptr);
        }
    }
}
